python/tchannel/tornado/peer.py

```diff
@@ -458,34 +506,79 @@ def send(self, arg1, arg2, arg3, traceflag=False, headers=None):
             parent_span_id = None
             trace_id = None
 
-        # set 'as' to 'raw if 'as' is empty
+        # set default transport headers
         headers = headers or {}
-        if 'as' not in headers:
-            headers['as'] = 'raw'
-
-        connection = yield self.peer.connect()
-        message_id = connection.next_message_id()
-        response = yield connection.send_request(
-            Request(
-                service=self.service,
-                argstreams=[InMemStream(endpoint), arg2, arg3],
-                id=message_id,
-                headers=headers,
-                tracing=Trace(
-                    name=endpoint,
-                    trace_id=trace_id,
-                    parent_span_id=parent_span_id,
-                    endpoint=Endpoint(self.peer.host,
-                                      self.peer.port,
-                                      self.service),
-                    traceflags=traceflag,
-                )
+        for k, v in self.headers.iteritems():
+            headers.setdefault(k, v)
+
+        peer = self.peer
+        connection = yield peer.connect()
+
+        request = Request(
+            service=self.service,
+            argstreams=[InMemStream(endpoint), arg2, arg3],
+            id=connection.next_message_id(),
+            headers=headers,
+            tracing=Trace(
+                name=endpoint,
+                trace_id=trace_id,
+                parent_span_id=parent_span_id,
+                endpoint=Endpoint(self.peer.host,
+                                  self.peer.port,
+                                  self.service),
+                traceflags=traceflag,
             )
         )
 
+        response = None
+        # black list to record all used peers, so they aren't chosen again.
+        blacklist = set()
+        # only retry on non-stream request
+        if request.is_streaming_request:
+            retry_times = 1
+            timeout_per_req = 0
+
+        # mac number of times to retry 3.
+        for num_of_retry in range(retry_times):
+            try:
+                response = yield self._retry_send(
+                    connection, request, timeout_per_req)
+                break
+            except (ProtocolError, TimeoutError) as e:
+                request.set_exception(e)
+                if not request.should_retry_on_error(e) or self._hostport:
```

Comment:
For `hostport` instead of checking here manually, you can do the same thing as `is_streaming_request` that you do above.